src/graph/ContentModel.ts

```diff
@@ -0,0 +1,208 @@
+import {
+  ModelNode,
+  ModelEdge,
+  ModelGraphData,
+  ModelAttributeSet,
+} from './types'
+import { v4 } from 'uuid'
+
+export class ContentModel {
```

Comment:
Could consider using immer to ensure immutability.
but maybe not necessary

Reply:
I thought about using an immutable library. The api is inspired by draftjs (which uses immutable.js). I decided not to use one straight away so I could get something off the ground